lib/matplotlib/testing/decorators.py

```diff
@@ -387,35 +387,45 @@ def decorator(func):
 
         _, result_dir = _image_directories(func)
 
-        if len(inspect.signature(func).parameters) == 2:
-            # Free-standing function.
-            @pytest.mark.parametrize("ext", extensions)
-            def wrapper(ext):
-                fig_test = plt.figure("test")
-                fig_ref = plt.figure("reference")
-                func(fig_test, fig_ref)
-                test_image_path = result_dir / (func.__name__ + "." + ext)
-                ref_image_path = (
-                    result_dir / (func.__name__ + "-expected." + ext))
-                fig_test.savefig(test_image_path)
-                fig_ref.savefig(ref_image_path)
-                _raise_on_image_difference(
-                    ref_image_path, test_image_path, tol=tol)
-
-        elif len(inspect.signature(func).parameters) == 3:
-            # Method.
-            @pytest.mark.parametrize("ext", extensions)
-            def wrapper(self, ext):
-                fig_test = plt.figure("test")
-                fig_ref = plt.figure("reference")
-                func(self, fig_test, fig_ref)
-                test_image_path = result_dir / (func.__name__ + "." + ext)
-                ref_image_path = (
-                    result_dir / (func.__name__ + "-expected." + ext))
-                fig_test.savefig(test_image_path)
-                fig_ref.savefig(ref_image_path)
-                _raise_on_image_difference(
-                    ref_image_path, test_image_path, tol=tol)
+        @pytest.mark.parametrize("ext", extensions)
+        def wrapper(*args, ext, **kwargs):
+            fig_test = plt.figure("test")
+            fig_ref = plt.figure("reference")
+            func(*args, fig_test=fig_test, fig_ref=fig_ref, **kwargs)
+            test_image_path = result_dir / (func.__name__ + "." + ext)
+            ref_image_path = result_dir / (
+                func.__name__ + "-expected." + ext
+            )
+            fig_test.savefig(test_image_path)
+            fig_ref.savefig(ref_image_path)
+            _raise_on_image_difference(
+                ref_image_path, test_image_path, tol=tol
+            )
+
+        sig = inspect.signature(func)
+        new_sig = sig.replace(
+            parameters=(
+                [
+                    param
+                    for param in sig.parameters.values()
+                    if param.name not in {"fig_test", "fig_ref"}
+                ]
+                + [
+                    inspect.Parameter(
+                        "ext", inspect.Parameter.POSITIONAL_OR_KEYWORD
+                    )
+                ]
+            )
+        )
+        wrapper.__signature__ = new_sig
+
+        # reach a bit into pytest internals to hoist the marks from
+        # our wrapped function
+        from _pytest.mark.structures import get_unpacked_marks
+
+        wrapper.pytestmark = get_unpacked_marks(
+            wrapper
+        ) + get_unpacked_marks(func)
```

Comment:
Is this not something you can add together from `wrapper.pytestmark` and `func.pytestmark`?

Reply:
Is that actually part of the public API?

Internally, pytest uses these three functions to do the mark manipulation:

```python
def get_unpacked_marks(obj):
    """
    obtain the unpacked marks that are stored on an object
    """
    mark_list = getattr(obj, "pytestmark", [])
    if not isinstance(mark_list, list):
        mark_list = [mark_list]
    return normalize_mark_list(mark_list)


def normalize_mark_list(mark_list):
    """
    normalizes marker decorating helpers to mark objects

    :type mark_list: List[Union[Mark, Markdecorator]]
    :rtype: List[Mark]
    """
    extracted = [
        getattr(mark, "mark", mark) for mark in mark_list
    ]  # unpack MarkDecorator
    for mark in extracted:
        if not isinstance(mark, Mark):
            raise TypeError("got {!r} instead of Mark".format(mark))
    return [x for x in extracted if isinstance(x, Mark)]


def store_mark(obj, mark):
    """store a Mark on an object
    this is used to implement the Mark declarations/decorators correctly
    """
    assert isinstance(mark, Mark), mark
    # always reassign name to avoid updating pytestmark
    # in a reference that was only borrowed
    obj.pytestmark = get_unpacked_marks(obj) + [mark]
```

I was between just vendoring them all, doing the semi-private import, or adding the `obj.pytestmark` list together.  I took the middle route, I'll change it to just adding the `pytestmark` lists together.